### PR TITLE
Add Nix build system

### PR DESCRIPTION
**Pull request by jakeisnt**
Nix is a system agnostic reproducible build system.

This will pin all of the dependencies, ensure we're all using the same versions of packages, etc. etc and provide an instant development environment. The only thing you have to have installed globally is Nix - not even node or yarn!

If you want to try this out:
- [Install Nix](https://nixos.org/guides/install-nix.html); it's one shell command
- enter the root directory of the project in your terminal
- run `nix-shell --run "yarn start"`

Bonus round: install [direnv](https://direnv.net/) to automatically load the dependencies!

### CODE REVIEW

**Comment by jakeisnt** [comment]
I also opted to fix the dependencies by removing '^' from each; that can be dangerous as packages can upgrade and randomly break or introduce security vulnerabilities. We should move many dependencies to devDependencies to decrease our build sizes but that's not a super pressing issue.

Try it out!

Notes:
- This assumes we will not be adding new packages frequently. To add a new package, you have to remove everything nix-generated, regenerate the yarn lockfile, then re-run `nix-shell`. This can be annoying.
- You do not have to participate in the Nix infrastructure if you don't want to; this can be used with your current nodejs/yarn setup already. Nix is something I enjoy and want to see in more projects as I think it provides a nice development experience.

**Comment by gamburgm** [comment]
I don't think adding Nix is a good idea here--I think it's unnecessary complexity. I have certainly had issues configuring Nix with the node ecosystem in the past, and even if that's unlikely or 'won't happen', if something unexpected does occur it's not going to be straightforward to resolve (mostly due to minimal adoption and instability of the project). I also suspect nobody will actually use Nix, and when whoever is using it leaves the team, it will become dead code that no one knows what to do with. I'd like to keep the development environment for the project to one build system, one way of doing things.

TL:DR; I'm cool with you using Nix locally, but it shouldn't be in the repo

**Comment by jakeisnt** [comment]
That's fair. I love using a single command to set up a development environment with Nix - projects that support `nix-shell` are so easy to get started on because there are no setup instructions or installation concerns; only the person who writes the nix configuration has to worry about how the dependencies are built (not sure if you read the code here, but this parses 'package.json' and 'yarn.lock' to install dependencies for the shell in effectively a container). 

That said, I understand that it's another tool that introduces additional complexity. I do think Nix is much better than VMs or containerized solutions like docker (docker can't ensure reproducible builds, is a royal pain to install, and containerization isn't /really/ necessary for local development) but for this project it isn't necessary.
